File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> ans = new ArrayList<Integer>();
    	if(matrix.length == 0 || matrix[0].length == 0) return ans;
            
        int minR = 0;
		int maxR = matrix.length-1;
		int minC = 0;
		int maxC = matrix[0].length-1;
		int count=0;
		
		while(count < matrix.length*matrix[0].length)
		{
			for(int i=minC;i<=maxC && count < matrix.length*matrix[0].length;i++)
			{
				ans.add(matrix[minR][i]);
				count++;
			}
			minR++;
			for(int i=minR;i<=maxR && count < matrix.length*matrix[0].length;i++)
			{
				ans.add(matrix[i][maxC]);
				count++;
			}
			maxC--;
			for(int i=maxC;i>=minC && count < matrix.length*matrix[0].length;i--)
			{
				ans.add(matrix[maxR][i]);
				count++;
			}
			maxR--;
			for(int i=maxR;i>=minR && count < matrix.length*matrix[0].length;i--)
			{
				ans.add(matrix[i][minC]);
				count++;
			}
			minC++;
		}
		return ans;
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        rows = len(matrix)
        cols = len(matrix[0])
        result = []
        
        if rows == 0 or cols == 0:
            return result

        rowmin = 0
        rowmax = rows - 1 
        colmin = 0
        colmax = cols - 1
        count = 0

        while count < (rows * cols):
            i = colmin
            while i <= colmax and count < (rows * cols):
                result.append(matrix[rowmin][i])
                count+=1
                i+=1
            rowmin+=1

            j = rowmin
            while j <= rowmax and count < (rows * cols):
                result.append(matrix[j][colmax])
                count+=1
                j+=1
            colmax-=1

            k = colmax
            while k >= colmin and count < (rows*cols):
                result.append(matrix[rowmax][k])
                count+=1
                k-=1
            rowmax-=1

            l = rowmax
            while l >= rowmin and count < (rows*cols):
                result.append(matrix[l][colmin])        
                count+=1
                l-=1
            colmin+=1
        return result